src/train.py

```diff
@@ -16,67 +16,248 @@
-            character, mel, mel_input, pos_text, pos_mel, text_len = data
-            character = character.to(DEVICE)
-            mel = mel.to(DEVICE)
-            enc_text, text_hidden_state, text_pad_mask = text_model.encode(character)
-            enc_speech, speech_hidden_state, speech_pad_mask = speech_model.encode(mel)
+        losses = defaultdict(list)
+        per, n_iters = 0, 0
+        
+        for batch in valid_dataset:
+            character, mel, mel_input, pos_text, pos_mel, text_len = batch
+
+            t_ae_loss, s_ae_loss = autoencoder_step(model, batch)
+            losses['t_ae'].append(t_ae_loss.detach().cpu().item())
+            losses['s_ae'].append(s_ae_loss.detach().cpu().item())
+
+            asr_loss, tts_loss = supervised_step(model, batch)
+            losses['asr_'].append(asr_loss.detach().cpu().item())
+            losses['tts_'].append(tts_loss.detach().cpu().item())
+
+            t_cm_loss, s_cm_loss = crossmodel_step(model, batch)
+            losses['s_cm'].append(s_cm_loss.detach().cpu().item())
+            losses['t_cm'].append(t_cm_loss.detach().cpu().item())
+
+            text_pred = model.asr(None, mel.to(DEVICE), infer=True).squeeze()
+            len_mask_max, len_mask_idx = torch.max((text_pred == PAD_IDX), dim=1)
+            len_mask_idx[end_mask_max == 0] = text_pred.shape[1] - 1
+            print("Lengths", len_mask_idx.shape)
+            per += compute_per(character.to(DEVICE), text_pred, text_len.to(DEVICE), len_mask_idx)
+            n_iters += 1
+
+    model.train()
+
+    return per/n_iters, losses
+
+
+def optimizer_step(loss, model, optimizer, args):
+
+    # Take a optimizer step!
+    optimizer.zero_grad()
+    loss.backward()
+    if args.grad_clip > 0.0:
+        nn.utils.clip_grad_norm_(model.parameters(), args.grad_clip)
+    optimizer.step()
+    return loss.detach().cpu().item()
+
+
+def autoencoder_step(model, batch):
+    """
+    Compute and return the loss for autoencoders
+    """
+    character, mel, mel_input, pos_text, pos_mel, text_len = batch
+    character, mel, mel_input = character.to(DEVICE), mel.to(DEVICE), mel_input.to(DEVICE)
+    mel_gold, char_gold = mel.detach(), character.detach().permute(1, 0)
+
+    text_pred = model.text_ae(character).permute(1, 2, 0)
+    pred, stop_pred = model.speech_ae(mel, mel_input)
+    
+    # TODO: add end length loss
+    s_ae_loss = speech_loss(mel_gold, pos_mel, pred, stop_pred)
+    t_ae_loss = text_loss(gold_char, pred_char)
+    return t_ae_loss, s_ae_loss
+
+
+def supervised_step(model, batch):
+    character, mel, mel_input, pos_text, pos_mel, text_len = batch
+    character, mel, mel_input = character.to(DEVICE), mel.to(DEVICE), mel_input.to(DEVICE)
+    mel_gold, char_gold = mel.detach(), character.detach().permute(1, 0)
+
+    pred, stop_pred = model.tts(character, mel_input)
+    text_pred = model.asr(character, mel).permute(1, 2, 0)
+    
+    tts_loss = speech_loss(mel_gold, pos_mel, pred, stop_pred)
```

Comment:
Referencing the previous comments, need to return the losses here and maybe weigh them differently.